proposals/module-types/Example-SharedEverythingDynamicLinking.md

```diff
@@ -0,0 +1,525 @@
+  (func (export "malloc") (param i32) (result i32) ...impl)
+  ...
+)
+```
+
+The SDK also contains standard library headers which contain declarations
+compatible with `libc.wasm`. To implement them, we first need some helper
+macros, which we'll conveniently define in `stddef.h`:
+
+```c
+/* stddef.h */
+#define WASM_IMPORT(module,name) __attribute__((import_module(#module), import_name(#name))) name
+#define WASM_EXPORT(name) __attribute__((export_name(#name))) name
+```
+
+These macros use the clang-specific attributes [`import_module`], [`import_name`]
+and [`export_name`] to ensure that the annotated C functions produce the
+correct wasm import or export definitions. Other compilers would use their own
+magic syntax to achieve the same effect. Using these macros, we can declare
+`malloc` in `stdlib.h`:
+
+```c
+/* stdlib.h */
+#include <stddef.h>
+#define LIBC(name) WASM_IMPORT(libc, name)
+void* LIBC(malloc)(size_t n);
+```
+
+## libzip
+
+With `libc` sketched out, we have a functional SDK we can now use to create the
+shared library `libzip`. Its C definition has the rough shape:
+
+```c
+/* libzip.h */
+#include <stddef.h>
+#define LIBZIP(name) WASM_IMPORT(libzip, name)
+void* LIBZIP(zip)(void* in, size_t in_size, size_t* out_size);
+```
+```c
+/* libzip.c */
+#include <stdlib.h>
+void* WASM_EXPORT(zip)(void* in, size_t in_size, size_t* out_size) {
+  ...
+  void *p = malloc(n);
+  ...
+}
+```
+
+Note that `libzip.h` annotates the `zip` declaration with an *import* attribute
+so that client modules generate proper wasm *import definitions* while `libzip.c`
+annotates the `zip` definition with an *export* attribute so this function generates
+a proper wasm *export definition*. Compiling with `clang -shared libzip.c`
+produces a `libzip.wasm` of the form:
+
+```wasm
+;; libzip.wat
+(module
+  (import "libc" (instance $libc
+    (export "memory" (memory 1))
+    (export "malloc" (func $malloc (param i32) (result i32)))
+  ))
+  (func (export "zip") (param i32 i32 i32) (result i32)
+    ...
+    call $libc.$malloc
+    ...
+  )
+)
+```
+
+## zipper
+
+Using `libc` and `libzip`, we can now write the `zipper` program:
+
+```c
+/* zipper.c */
+#include <stdlib.h>
+#include "libzip.h"
+int main(int argc, char* argv[]) {
+  ...
+  void *in = malloc(n);
+  ...
+  void *out = zip(in, n, &out_size);
+  ...
+}
+```
+
+When compiled with `clang zipper.c`, we get a `zipper.wasm` shaped like:
+
+```wasm
+;; zipper.wat
+(module
+  (type $Libc (instance
+    (export "memory" (memory 1))
+    (export "malloc" (func $malloc (param i32) (result i32)))
+  ))
+  (type $LibZip (instance
+    (export "zip" (func $zip (param i32 i32 i32) (result i32)))
+  ))
+
+  (import "libc" (module $LIBC
+    (exports $Libc)
+  ))
+  (import "libzip" (module $LIBZIP
+    (import "libc" (instance (type $Libc)))
+    (exports $LibZip)
+  ))
+
+  (instance $libc (instance.instantiate $LIBC))
+  (instance $libzip (instance.instantiate $LIBZIP (ref.instance $libc)))
```

Comment:
One thing I'm curious about here is the conventions with the toolchain. When generating shared libraries all of the imports are of instances (right?) but here for binaries (zipper.c and imgmgk.c) you've got the binary compilations importing modules and then doing all of the instantiations.

Although this definitely seems valid given this explainer, this touches on the convention of "who's responsible for creating all the instances?" At least for this shared-everything case I think the answer is definitely the top-level module, but I'm not sure that the native C toolchain (clang/lld/rustc/etc) should be responsible for generating the top-level module.

I think it might be more consistent if the output of the toolchain is always "here's the instances I import, and given that functionality I give you this functionality". That I think is a bit more conventional with how compilers are set up today because you wouldn't need so much contextual information when linking (you'd have to transitively read off all dependencies and such). That I think also sort of mirrors how native object file formats work where a dynamic library only lists its dependencies, not its transitive dependencies (I think?)

Now that being said I think there's definitely still be a standard tool which wraps everything up. That'd be shared across language ecosystems and would be effectively be `ld.so`, only at compile time but not runtime. That'd have the job of "given this search path for wasm files as well as this main wasm file, produce a new wasm file which imports modules instead of instances and does all the instantiation internally".

Reply:
For the C/C++ toolchain, I think it's a bit more strictly defined than that.  I think there are two cases:
* compiling with `-shared`: all shared libraries are imported as instances
* compiling without shared (with a `main()` or new-typed-main): instances are created privately

The only interesting choice is when do module imports of shared libraries get "bundled" (into nested modules), if ever.  As the explainer says, that's a question for a higher-level tool that's able to see a whole application.